File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.configuration.base;

import org.kaaproject.kaa.client.KaaClientProperties;
import org.kaaproject.kaa.client.configuration.ConfigurationHashContainer;
import org.kaaproject.kaa.client.configuration.ConfigurationProcessor;
import org.kaaproject.kaa.client.configuration.storage.ConfigurationStorage;
import org.kaaproject.kaa.client.context.ExecutorContext;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.common.hash.EndpointObjectHash;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Collections;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public abstract class AbstractConfigurationManager implements ConfigurationManager {
  private static final Logger LOG = LoggerFactory.getLogger(AbstractConfigurationManager.class);
  protected final ConfigurationDeserializer deserializer;
  private final Set<ConfigurationListener> listeners = Collections.newSetFromMap(
          new ConcurrentHashMap<ConfigurationListener, Boolean>());
  private final KaaClientProperties properties;
  private final ExecutorContext executorContext;
  private volatile byte[] configurationData;
  private ConfigurationStorage storage;
  private ConfigurationHashContainer container = new HashContainer();
  private KaaClientState state;

  /**
   * All-args constructor.
   */
  public AbstractConfigurationManager(KaaClientProperties properties, KaaClientState state,
                                      ExecutorContext executorContext) {
    super();
    this.properties = properties;
    this.state = state;
    this.executorContext = executorContext;
    this.deserializer = new ConfigurationDeserializer(executorContext);
  }

  private static byte[] toByteArray(ByteBuffer buffer) {
    if (buffer == null) {
      return null;
    }
    byte[] bytes = new byte[buffer.remaining()];
    buffer.get(bytes);
    return bytes;
  }

  @Override
  public void init() {
    getConfigurationData();
  }

  @Override
  public boolean addListener(ConfigurationListener listener) {
    if (listener != null) {
      LOG.trace("Adding listener {}", listener);
      return listeners.add(listener);
    } else {
      throw new RuntimeException("Can't add null as a listener");
    }
  }

  @Override
  public boolean removeListener(ConfigurationListener listener) {
    if (listener != null) {
      return listeners.remove(listener);
    } else {
      throw new RuntimeException("Can't remove null listener");
    }
  }

  @Override
  public ConfigurationProcessor getConfigurationProcessor() {
    return new ConfigurationProcessor() {

      @Override
      public void processConfigurationData(ByteBuffer buffer, boolean fullResync)
              throws IOException {
        if (fullResync) {
          configurationData = toByteArray(buffer);
          if (LOG.isTraceEnabled()) {
            LOG.trace("Received configuration data {}", Arrays.toString(configurationData));
          }
          if (storage != null) {
            LOG.debug("Persisting configuration data from storage {}", storage);
            storage.saveConfiguration(ByteBuffer.wrap(configurationData));
            LOG.debug("Persisted configuration data from storage {}", storage);
          }
          deserializer.notify(Collections.unmodifiableCollection(listeners), configurationData);
        } else {
          LOG.warn("Only full resync delta is supported!");
        }
      }
    };
  }

  @Override
  public void setConfigurationStorage(ConfigurationStorage storage) {
    this.storage = storage;
  }

  @Override
  public ConfigurationHashContainer getConfigurationHashContainer() {
    return container;
  }

  protected byte[] getConfigurationData() {
    if (configurationData == null) {
      configurationData = loadConfigurationData();
    }
    return configurationData;
  }

  private byte[] loadConfigurationData() {
    if (storage != null) {
      if (state.isConfigurationVersionUpdated()) {
        LOG.info("Clearing old configuration data from storage {}", storage);
        try {
          storage.clearConfiguration();
        } catch (IOException ex) {
          LOG.error("Failed to clear configuration from storage", ex);
        }
      } else {
        LOG.debug("Loading configuration data from storage {}", storage);
        try {
          configurationData = toByteArray(storage.loadConfiguration());
        } catch (IOException ex) {
          LOG.error("Failed to load configuration from storage", ex);
        }
      }
    }
    if (configurationData == null) {
      LOG.debug("Loading configuration data from defaults {}", storage);
      configurationData = getDefaultConfigurationData();
    }
    if (LOG.isTraceEnabled()) {
      LOG.trace("Loaded configuration data {}", Arrays.toString(configurationData));
    }
    return configurationData;
  }

  protected byte[] getDefaultConfigurationData() {
    return properties.getDefaultConfigData();
  }

  private class HashContainer implements ConfigurationHashContainer {
    @Override
    public EndpointObjectHash getConfigurationHash() {
      return EndpointObjectHash.fromSha1(getConfigurationData());
    }
  }
}

File: client/client-multi/client-objective-c/Kaa/configuration/AbstractConfigurationManager.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "AbstractConfigurationManager.h"
#import "KaaLogging.h"

#pragma clang diagnostic ignored "-Wprotocol"

#define TAG @"AbstractConfigurationManager >>>"

@interface AbstractConfigurationManager ()

@property(nonatomic, strong) KaaClientProperties *properties;
@property(nonatomic, strong) id<KaaClientState> state;
@property(nonatomic, strong) id<ConfigurationStorage> storage;
@property(nonatomic, strong) NSData *configurationData;
@property(nonatomic, strong) id<ExecutorContext> executorContext;

@property(strong) NSMutableSet *delegates;
@property(strong) NSLock *delegatesLock;

- (NSData *)loadConfigurationData;

@end

@implementation AbstractConfigurationManager

- (instancetype)initWithClientProperties:(KaaClientProperties *)properties
                                   state:(id<KaaClientState>)state
                         executorContext:(id<ExecutorContext>)context {
    self = [super init];
    if (self) {
        self.delegates = [NSMutableSet set];
        self.properties = properties;
        self.state = state;
        _deserializer = [[ConfigurationDeserializer alloc] initWithExecutorContext:context];
        self.executorContext = context;
    }
    return self;
}

- (void)initiate {
    [self getConfigurationData];
    DDLogDebug(@"%@ Configuration manager init completed!", TAG);
}

- (void)addDelegate:(id<ConfigurationDelegate>)delegate {
    if (delegate) {
        DDLogVerbose(@"%@ Adding delegate %@", TAG, delegate);
        [self.delegatesLock lock];
        [self.delegates addObject:delegate];
        [self.delegatesLock unlock];
    } else {
        [NSException raise:NSInvalidArgumentException format:@"Can't add nil delegate"];
    }
}

- (void)removeDelegate:(id<ConfigurationDelegate>)delegate {
    if (delegate) {
        DDLogVerbose(@"%@ Removing delegate", TAG);
        [self.delegatesLock lock];
        [self.delegates removeObject:delegate];
        [self.delegatesLock unlock];
    } else {
        [NSException raise:NSInvalidArgumentException format:@"Can't remove nil delegate"];

    }
}

- (void)processConfigurationData:(NSData *)data fullResync:(BOOL)fullResync {
    if (fullResync) {
        self.configurationData = data;
        DDLogVerbose(@"%@ Received configuration data: %@", TAG, self.configurationData);
        if (self.storage) {
            DDLogDebug(@"%@ Persisting configuration data from storage: %@", TAG, self.storage);
            [self.storage saveConfiguration:self.configurationData];
            DDLogDebug(@"%@ Persisted configuration data from storage: %@", TAG, self.storage);
        }
        [self.delegatesLock lock];
        [_deserializer notifyDelegates:self.delegates withData:self.configurationData];
        [self.delegatesLock unlock];
    } else {
        DDLogWarn(@"%@ Only full resync delta is supported!", TAG);
    }
}

- (EndpointObjectHash *)getConfigurationHash {
    return [EndpointObjectHash hashWithSHA1:[self getConfigurationData]];
}

- (void)setConfigurationStorage:(id<ConfigurationStorage>)storage {
    self.storage = storage;
}

- (NSData *)loadConfigurationData {
    if (self.storage) {
        if ([self.state isConfigurationVersionUpdated]) {
            DDLogDebug(@"%@ Clearing old configuration data from storage: %@", TAG, self.storage);
            @try {
                [self.storage clearConfiguration];
            }
            @catch (NSException *exception) {
                DDLogError(@"%@ Failed to clear configuration from storage: %@", TAG, exception);
            }
        } else {
            DDLogDebug(@"%@ Loading configuration data from storage: %@", TAG, self.storage);
            @try {
                self.configurationData = [self.storage loadConfiguration];
            }
            @catch (NSException *exception) {
                DDLogError(@"%@ Failed to load configuration from storage: %@", TAG, exception);
            }
        }
    }
    if (!self.configurationData) {
        DDLogDebug(@"%@ Loading configuration data from defaults: %@", TAG, self.storage);
        self.configurationData = [self getDefaultConfigurationData];
    }
    DDLogVerbose(@"%@ Loaded configuration data: %@", TAG, self.configurationData);
    return self.configurationData;
}

- (NSData *)getConfigurationData {
    if (!self.configurationData) {
        self.configurationData = [self loadConfigurationData];
    }
    return self.configurationData;
}

- (NSData *)getDefaultConfigurationData {
    return [self.properties defaultConfigData];
}

@end
